### PR TITLE
[Qwen3TTS][Perf] Add CodePredictorGraph: capture full AR loop as single CUDA graph

### PR DESCRIPTION
**Pull request by linyueqian**
## Purpose

Add `CodePredictorGraph` to capture the Qwen3-TTS code predictor's full 16-step autoregressive loop as a single CUDA graph, replacing the eager multi-step forward pass (dynamic tensor shapes + `torch.multinomial`).

**Key changes:**
- `qwen3_tts_code_predictor_vllm.py`: New `CodePredictorGraph` class using static KV caches (`[B, kv_heads, 16, head_dim]`), pre-built causal masks, and `torch.argmax` (graph-safe). Captures all 16 AR steps as one `torch.cuda.CUDAGraph`. Supports multiple batch sizes — one graph captured per batch size, smallest-fit dispatch with input padding.
- `gpu_model_runner.py`: Call `code_predictor.setup_graph()` in `OmniGPUModelRunner.load_model()` when CUDA graphs are enabled. TTS-only (Omni nests code predictor under `.talker` and is handled separately).

**Trade-off**: graph path uses `argmax` (greedy/deterministic); eager fallback retains `multinomial` sampling for batch > max captured size.

## Test Plan

```bash
# End-to-end correctness
python examples/offline_inference/qwen3_tts/end2end.py -q CustomVoice --stage-init-timeout 600

# bs=1 latency benchmark (CUDA graph vs eager)
python examples/offline_inference/qwen3_tts/benchmark_cudagraph.py --num-prompts 6 --warmup 1 --mode both
```

## Test Result

**bs=1 isolated code predictor benchmark** (A100-80GB):

| Mode | Mean latency | Speedup |
|------|-------------|---------|
| CUDA graph | 1.43s | **3.54x** |
| Eager | 5.07s | 1.0x |

**End-to-end throughput across batch sizes** (includes Code2Wav Stage 1, always eager):

| Mode | bs=1 | bs=4 |
|------|------|------|
| CUDA graph | 0.29 req/s | **0.62 req/s** |
| Eager | 0.17 req/s | 0.45 req/s |

benchmark_batch_results

### CODE REVIEW

**Comment by hsliuustc0106** [comment]
## Summary (2-3 sentences)
This PR adds `CodePredictorGraph` to capture Qwen3-TTS's 16-step autoregressive loop as a single CUDA graph, replacing eager execution. The implementation uses static KV caches, pre-built masks, and argmax sampling for graph compatibility.

## Red Flags

- **Missing tests:** Yes - no test coverage for CUDA graph capture, fallback logic, or multi-batch-size dispatch
- **Missing error handling:** Partial - graph capture failures not explicitly handled
- **Breaking changes:** No - adds new code path, eager fallback preserved
- **Security concerns:** No

## Pros

- Follows established pattern from other CUDA graph implementations in vllm
- Preserves eager fallback for batch sizes exceeding captured graphs (qwen3_tts_code_predictor_vllm.py:450-455)
- Static KV cache design avoids dynamic tensor shapes (qwen3_tts_code_predictor_vllm.py:380-385)
- Integration point is clean (gpu_model_runner.py:285-290)

## Cons

- No test coverage for CUDA graph functionality - missing tests for capture, replay, fallback, and multi-batch dispatch
- Trade-off documentation incomplete (qwen3_tts_code_predictor_vllm.py:15-20) - doesn't quantify when argmax vs multinomial matters
- Missing error handling for graph capture failures (qwen3_tts_code_predictor_vllm.py:420-430)
- No validation that 16-step loop assumption holds for all Qwen3-TTS configs
- Test plan in PR body shows manual testing only - no automated regression tests added
- Graph warmup cost not documented - how many batch sizes are captured? What's the memory overhead?

## Verdict

Request changes - need tests and performance measurements before merge.

File: vllm_omni/worker/gpu_model_runner.py [hunk 1]
```diff
@@ -95,6 +96,18 @@ def load_model(self, *args, **kwargs) -> None:
             )
             self.last_talker_hidden = self._make_buffer(max_batch_size, hidden_size, dtype=self.dtype, numpy=False)
             self.text_step = self._make_buffer(max_batch_size, hidden_size, dtype=self.dtype, numpy=False)
+            # Set up CUDA graph for TTS code predictor.
+            # TTS models expose .code_predictor directly; Omni nests it under .talker.
+            code_predictor = getattr(self.model, "code_predictor", None)
+            if code_predictor is not None and cudagraph_mode.has_full_cudagraphs():
+                import math
+
+                bs_cap = min(max(self.max_num_reqs, 1), 8)
```
Comment (hsliuustc0106) [inline vllm_omni/worker/gpu_model_runner.py]: could this be moved out of model_runner?

File: vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py [hunk 2]
```diff
@@ -314,182 +697,119 @@ def __init__(
         else:
             self.small_to_mtp_projection = nn.Identity()
 
-        self._kv_cache: _LocalPredictorKVCache | None = None
+        # Sampling parameters (temperature + top_k).
+        self.temperature = 0.9
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py]: why this is fixed inside model definition?

File: vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py [hunk 3]
```diff
@@ -314,182 +697,119 @@ def __init__(
...
-        # Flatten to token-major.
-        hs = inputs_embeds.to(dtype=torch.bfloat16).reshape(bsz * qlen, -1)
-        hs = self.small_to_mtp_projection(hs)
-
-        query_lens = torch.full((bsz,), qlen, dtype=torch.int32)
-        seq_lens = query_lens.clone()
-        attn_metadata, positions, slot_mappings = self._kv_cache.build_attn_metadata(
-            num_reqs=bsz, query_lens=query_lens, seq_lens=seq_lens
-        )
+    def setup_graph(self, batch_sizes: tuple[int, ...] = (1,)) -> None:
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py]: Missing test coverage for CUDA graph capture. Add tests that verify: (1) graph captures successfully for different batch sizes, (2) captured graphs produce correct outputs, (3) fallback to eager mode works when batch size exceeds captured sizes. Without these tests, we can't validate the graph functionality or prevent regressions.

File: vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py [hunk 4]
```diff
@@ -279,8 +393,279 @@ def load_weights(self, weights: Iterable[tuple[str, torch.Tensor]]) -> set[str]:
...
+class CodePredictorGraph:
+    """Captures the full code predictor AR loop as a single CUDA graph.
+
+    Replaces the eager multi-step forward (dynamic shapes + torch.multinomial)
+    with a static, graph-safe version using:
+    - Pre-allocated KV caches (fixed shapes, written at static Python-int offsets)
+    - torch.argmax instead of torch.multinomial (deterministic, graph-safe)
+    - All steps unrolled inside a single torch.cuda.CUDAGraph context
+
+    NOTE: argmax is greedy/deterministic; the eager fallback uses multinomial
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py:414]: Where are the measurements quantifying the argmax vs multinomial trade-off? The comment states argmax is deterministic but doesn't explain when this matters for TTS quality. Provide perceptual quality metrics (MOS scores, similarity tests) comparing argmax and multinomial sampling on realistic workloads.

File: vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py [hunk 5]
```diff
@@ -279,8 +393,279 @@ def load_weights(self, weights: Iterable[tuple[str, torch.Tensor]]) -> set[str]:
...
+                self._reset_kv_caches()
+                self._full_loop()
+        torch.cuda.current_stream().wait_stream(warmup_stream)
+
+        gc.collect()
+        torch.cuda.empty_cache()
+
+        self._graph = torch.cuda.CUDAGraph()
+        self._reset_kv_caches()
+        with torch.cuda.graph(self._graph):
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py]: No error handling for graph capture failures. What happens if CUDA graph capture fails due to memory constraints or unsupported operations? Add try-except around graph capture with graceful fallback to eager mode and logging of the failure reason.

File: vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py [hunk 6]
```diff
@@ -314,182 +697,119 @@ def __init__(
...
         last_talker_hidden: torch.Tensor,
-        do_sample: bool = True,
-        temperature: float = 1.0,
-        top_k: int = 50,
-        top_p: float = 1.0,
     ) -> torch.Tensor:
         """Full autoregressive prediction of residual codebooks 1..Q-1.
 
+        Dispatches to CodePredictorGraph (argmax, CUDA graph) for batch_size=1
+        when a graph has been captured via setup_graph(). Falls back to the
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py]: Missing validation that the 16-step assumption holds for all Qwen3-TTS configurations. Add assertion or config validation to ensure num_ar_steps==16, or make the implementation flexible to handle different step counts. Hard-coding 16 steps could break with future model variants.

**Comment by hsliuustc0106** [comment]
**vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py:735**

Missing test coverage for CUDA graph capture. Add tests that verify: (1) graph captures successfully for different batch sizes, (2) captured graphs produce correct outputs, (3) fallback to eager mode works when batch size exceeds captured sizes. Without these tests, we can't validate the graph functionality or prevent regressions.

---
*Note: Could not post as inline comment. Error: gh: Validation Failed (HTTP 422)*

**Comment by hsliuustc0106** [comment]
**vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py:410**

Where are the measurements quantifying the argmax vs multinomial trade-off? The comment states argmax is deterministic but doesn't explain when this matters for TTS quality. Provide perceptual quality metrics (MOS scores, similarity tests) comparing argmax and multinomial sampling on realistic workloads.

---
*Note: Could not post as inline comment. Error: gh: Validation Failed (HTTP 422)*

File: vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py [hunk 7]
```diff
@@ -279,8 +393,279 @@ def load_weights(self, weights: Iterable[tuple[str, torch.Tensor]]) -> set[str]:
...
+        gc.collect()
+        torch.cuda.empty_cache()
+
+        self._graph = torch.cuda.CUDAGraph()
+        self._reset_kv_caches()
+        with torch.cuda.graph(self._graph):
+            self._full_loop()
+
+        logger.info("CodePredictorGraph: capture complete.")
+
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py:642]: No error handling for graph capture failures. What happens if CUDA graph capture fails due to memory constraints or unsupported operations? Add try-except around graph capture with graceful fallback to eager mode and logging of the failure reason.

**Comment by hsliuustc0106** [comment]
**vllm_omni/worker/gpu_model_runner.py:110**

Where are the performance benchmarks showing this optimization's impact? The PR claims performance improvement but provides no latency or throughput measurements. Add benchmark results comparing eager vs graph execution for typical batch sizes (1, 2, 4, 8) with realistic inputs.

---
*Note: Could not post as inline comment. Error: gh: Validation Failed (HTTP 422)*

File: vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py [hunk 8]
```diff
@@ -314,182 +697,119 @@ def __init__(
...
-        top_p: float = 1.0,
     ) -> torch.Tensor:
         """Full autoregressive prediction of residual codebooks 1..Q-1.
 
+        Dispatches to CodePredictorGraph (argmax, CUDA graph) for batch_size=1
+        when a graph has been captured via setup_graph(). Falls back to the
+        original eager path (multinomial sampling) otherwise.
+
         Args:
             layer0_code: [B, 1] first-layer codec token ids.
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py:800]: Missing validation that the 16-step assumption holds for all Qwen3-TTS configurations. Add assertion or config validation to ensure num_ar_steps==16, or make the implementation flexible to handle different step counts. Hard-coding 16 steps could break with future model variants.

File: vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py (hunk 3 above)
Comment (linyueqian) [reply vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py]: tested locally. i can add test if needed. see the figure above.

File: vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py (hunk 6 above)
Comment (linyueqian) [reply vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py]: nothing is hardcoded. also all three variants use num_code_groups=16 with the same 12Hz tokenizer.

File: vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py [hunk 9]
```diff
@@ -314,182 +719,140 @@ def __init__(
...
 
-            params = dict(self.named_parameters(remove_duplicate=False))
-            for name, w in other_weights:
-                if name not in params:
-                    continue
-                default_weight_loader(params[name], w)
-                loaded.add(name)
-            return loaded
+        params = dict(self.named_parameters(remove_duplicate=False))
+        for name, w in other_weights:
```
Comment (lishunyang12) [inline vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_code_predictor_vllm.py:749]: The eager fallback now does full recompute every step (growing `current_input` by concatenation, re-running the entire sequence through the transformer each iteration). The old code used incremental KV-cached decode (`prefill_logits` + `decode_logits`), which was O(n) per step. This makes the fallback path O(n^2) in total sequence length — a significant regression for batch sizes that exceed the captured graph sizes.

File: vllm_omni/core/sched/omni_generation_scheduler.py [hunk 10]
```diff
@@ -451,7 +451,7 @@ def update_from_output(
                         events=request.take_events(),
                         kv_transfer_params=kv_transfer_params,
                         trace_headers=request.trace_headers,
-                        num_cached_tokens=request.num_cached_tokens,
+                        num_cached_tokens=max(0, request.num_cached_tokens),
```
Comment (lishunyang12) [inline vllm_omni/core/sched/omni_generation_scheduler.py:454]: Unrelated bugfix — should this be a separate PR? If not, at least add a comment explaining when `num_cached_tokens` can go negative.

**Comment by linyueqian** [comment]
@lishunyang12 this pr should be closed. sorry for not doing that in time.

File: vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_talker.py [hunk 11]
```diff
@@ -1532,6 +1521,13 @@ def _talker_and_collect_speaker(ws: Iterable[tuple[str, torch.Tensor]]):
...
+
+        cudagraph_mode = self.vllm_config.compilation_config.cudagraph_mode
+        if cudagraph_mode is not None and cudagraph_mode.has_full_cudagraphs():
+            max_num_reqs = self.vllm_config.scheduler_config.max_num_seqs
+            bs_cap = min(max(max_num_reqs, 1), 8)
+            self.code_predictor.setup_graph(max_batch_size=bs_cap)
+
         return loaded
 
     # -------------------- GPU-side MTP fast-path --------------------
```
Comment (lishunyang12) [inline vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_talker.py:1533]: Calling setup_graph() inside load_weights() is surprising — graph capture is a runtime concern, not a weight-loading concern. Why not call it from OmniGPUModelRunner.load_model() like the CUDAGraphWrapper setup for the talker?

**Review by lishunyang12 — COMMENTED** [review]
Left a couple comments. The CUDA graph capture approach looks solid — main concern is that the eager fallback path regressed from incremental KV-cached decode to full-recompute, which makes the fallback significantly slower than before this PR.
